crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -1711,6 +1659,84 @@ where
+                let result = if self.sync.is_pipeline_idle() {
+                    // we can only insert new payloads if the pipeline is _not_ running, because it
+                    // holds exclusive access to the database
+                    self.try_insert_new_payload(block)
+                } else {
+                    self.try_buffer_payload(block)
+                };
+
+                let status = match result {
+                    Ok(status) => status,
+                    Err(error) => {
+                        warn!(target: "consensus::engine", %error, "Error while processing payload");
+
+                        let (block, error) = error.split();
+                        if !error.is_invalid_block() {
+                            // TODO: revise if any error should be considered fatal at this point.
+                            let _ =
+                                tx.send(Err(BeaconOnNewPayloadError::Internal(Box::new(error))));
+                            return Ok(EngineEventOutcome::Processed)
+                        }
+
+                        // If the error was due to an invalid payload, the payload is added to the
+                        // invalid headers cache and `Ok` with [PayloadStatusEnum::Invalid] is
+                        // returned.
+                        warn!(target: "consensus::engine", invalid_hash=?block.hash(), invalid_number=?block.number, %error, "Invalid block error on new payload");
+                        let latest_valid_hash = self
+                            .latest_valid_hash_for_invalid_payload(block.parent_hash, Some(&error));
+                        // keep track of the invalid header
+                        self.invalid_headers.insert(block.header);
+                        PayloadStatus::new(
+                            PayloadStatusEnum::Invalid { validation_error: error.to_string() },
+                            latest_valid_hash,
+                        )
+                    }
+                };
+
+                if status.is_valid() {
+                    if let Some(target) = self.forkchoice_state_tracker.sync_target_state() {
+                        // if we're currently syncing and the inserted block is the targeted
+                        // FCU head block, we can try to make it canonical.
+                        if block_hash == target.head_block_hash {
+                            if let Err((_hash, error)) =
+                                self.try_make_sync_target_canonical(block_num_hash)
+                            {
+                                if error.is_fatal() {
+                                    let response = Err(BeaconOnNewPayloadError::Internal(
+                                        Box::new(error.clone()),
+                                    ));
+                                    let _ = tx.send(response);
+                                    return Err(RethError::Canonical(error))
```

Comment:
this introduces different behavior: previously the engine would not terminate upon encountering fatal error during new payload processing